Pass context to completion

lukasmalkmus opened the pull request:
Closes #1260.
Closes #1263.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/spf13/cobra?pullRequest=1265) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/spf13/cobra?pullRequest=1265) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/spf13/cobra?pullRequest=1265) it.</sub>

marckhouzam commented:
Thanks @lukasmalkmus. Could you give a snippet of code that shows how a context is set so I can test?  I've never used one but I assume it should be set on the root command?

lukasmalkmus commented:
Yes, use `rootCommand.ExecuteContext()`.

marckhouzam commented:
@lukasmalkmus I've never used go contexts before.  Could you provide a small test program to show how you would generate the context and then how you would use it?

lukasmalkmus commented:
I'll extend this example from the docs:https://github.com/spf13/cobra/blob/master/shell_completions.md#dynamic-completion-of-nouns

So imagine you got the linked command and want to do dynamic completion of nouns:

´´´go
cmd := &cobra.Command{
	Use:   "status RELEASE_NAME",
	Short: "Display the status of the named release",
	Long:  status_long,
	RunE: func(cmd *cobra.Command, args []string) {
		RunGet(args[0])
	},
	ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
		if len(args) != 0 {
			return nil, cobra.ShellCompDirectiveNoFileComp
		}
		return getReleasesFromCluster(toComplete), cobra.ShellCompDirectiveNoFileComp
	},
}
```

Now we might want to use a context in our application for various reasons (cancelation, passing metadata (gRPC), etc.):

```go
cmd := &cobra.Command{
	Use:   "status RELEASE_NAME",
	Short: "Display the status of the named release",
	Long:  status_long,
	RunE: func(cmd *cobra.Command, args []string) {
		RunGet(args[0])
	},
	ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
		if len(args) != 0 {
			return nil, cobra.ShellCompDirectiveNoFileComp
		}
		return getReleasesFromCluster(toComplete), cobra.ShellCompDirectiveNoFileComp
	},
}

func getReleasesFromCluster(ctx context.Context, toComplete string) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, "http://serve.com/api/releases", nil)
	if err != nil {
		return err
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}

	// ...
}
```

To actually inject the context, it must be passed using `cmd.ExecuteContext()`. However, this currently fails on master because the context isn't passed down into the `ValidArgsFunction`.

I also want to use the functionally of `cmd.ExecuteC()` but still need the context. That's why I added `cmd.ExecuteContextC()`.

Feel free to reach out if you need more details.

lukasmalkmus commented:
Just noticed I didn't add `ExecuteContextC` as requested in #1260 😅

lukasmalkmus commented:
And added it. This now closes #1260 as well.

jharshman commented:
@lukasmalkmus can you add some tests?

lukasmalkmus commented:
@jharshman Done! I updated the helper function `executeCommandWithContext` to use `ExecuteContextC` under the hood. The implementation of `ExecuteContextC` is ridiculously simple but using this change makes sure it gets some usage in tests.

I also added `TestValidArgsFuncCmdContext` which makes sure a valid context is received in a `ValidArgsFunc`.

lukasmalkmus commented:
Can someone unmark #1260 as being stale?

marckhouzam commented on custom_completions_test.go:
```diff
@@ -1194,6 +1195,44 @@ func TestFlagDirFilterCompletionInGo(t *testing.T) {
+	validArgsFunc := func(cmd *Command, args []string, toComplete string) ([]string, ShellCompDirective) {
+		if len(args) != 0 {
+			return nil, ShellCompDirectiveNoFileComp
+		}
```
Do we need this if statement?

lukasmalkmus replied:
I got rid of it since it doesn't affect the test. However, I copied that from the `validArgsFunc` at the top of the file. I guess there might be a reason for it.

marckhouzam reviewed (changes requested):
Thanks for this.  I added a couple of comments in-line.

marckhouzam commented on custom_completions_test.go:
```diff
@@ -1194,6 +1195,44 @@ func TestFlagDirFilterCompletionInGo(t *testing.T) {
 	}
 }
 
+func TestValidArgsFuncCmdContext(t *testing.T) {
```
There are two reasons for this problem:
1. the code in the `validArgsFunc` must extract the context using `cmd.Context()`
2. `executeCommandWithContext` should be called on a sub-command.  Calling it on the root command is not a sufficient test because the root command already has the context set.

lukasmalkmus replied:
Thanks, good catch!

I made sure 1. does not happen again. I also call completion on a child command now. removing `finalCmd.ctx = c.ctx` now correctly fails the test.

lukasmalkmus commented:
Will address soon, thanks for you in depth review. Should have spent some more time on that myself tbh.

lukasmalkmus commented on command_test.go:
```diff
@@ -20,15 +20,15 @@ func executeCommand(root *Command, args ...string) (output string, err error) {
 	root.SetArgs(args)
 
-	err = root.ExecuteContext(ctx)
+	c, err = root.ExecuteContextC(ctx)
```
Is that an actual problem? `ExecuteContext()` just calls `ExecuteC()` under the hood but discards the first return argument (the command).

lukasmalkmus replied:
Alright, got it.

I just reverted my initial changes to those methods and just replicated the helper method as well as the test for the new code path.

lukasmalkmus commented:
Took me some time, but should be ok now.

lukasmalkmus commented:
Test fails are not related to my changes. I just need https://github.com/spf13/cobra/pull/1309 :)

marckhouzam reviewed (changes requested):
Thanks @lukasmalkmus.

I explained better (I hope) why I think it is important to test `ExecuteContext()`.

The rest looks good.
Can you also rebase the PR to the latest master? #1309 has just been merged.

lukasmalkmus commented:
Resolved all discussions and rebased on latest master.

marckhouzam reviewed (approved):
To looks good to me.  Thanks @lukasmalkmus for addressing all comments.
One last thing is that usually I'm asked to squash all the commits into one before the PR is merged.

@jharshman I believe this one is ready.

lukasmalkmus commented:
✅

matanw commented:
It seems ready to merge, no? any reason no to merge it?

neilotoole commented:
Can we get this merged in plz?

jpmcb reviewed (approved):
This looks good to me! - and thanks for adding tests, makes it easier to review!
